File: wlxx-services/wlxx-system/wlxx-system-provider/src/main/java/com/xzg/wlxx/system/service/impl/OrgServiceImpl.java
package com.xzg.wlxx.system.service.impl;

import cn.hutool.core.bean.BeanUtil;
import com.baomidou.mybatisplus.core.toolkit.Wrappers;
import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl;
import com.xzg.wlxx.common.exception.BusinessException;
import com.xzg.wlxx.common.util.IteratorUtils;
import com.xzg.wlxx.common.util.PojoConvertor;
import com.xzg.wlxx.system.client.entity.po.OrgPo;
import com.xzg.wlxx.system.client.entity.vo.OrgVo;
import com.xzg.wlxx.system.mapper.OrgMapper;
import com.xzg.wlxx.system.service.OrgService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

@Service
@Slf4j
@RequiredArgsConstructor
public class OrgServiceImpl extends ServiceImpl<OrgMapper, OrgPo> implements OrgService {
    @Override
    public Boolean add(OrgPo org) {
        boolean isRepeat = lambdaQuery()
                .eq(OrgPo::getOrgCode, org.getOrgCode())
                .exists();
        if (isRepeat) {
            throw new BusinessException("orgCode is repeat");
        }
        return save(org);
    }

    @Override
    public List<OrgVo> list4Tree(Long rootId) {
        List<OrgPo> list = baseMapper.selectList(Wrappers.<OrgPo>lambdaQuery()
                .orderByAsc(OrgPo::getOrgLevel));
        Integer rootLevel = list.stream().findFirst().get().getOrgLevel();
        List<OrgVo> result = list.stream()
                .filter(e ->
                        Objects.nonNull(rootId)
                                ? Objects.equals(e.getId(), rootId)
                                : Objects.equals(e.getOrgLevel(), rootLevel
                        )
                )
                .map(e -> {
                    OrgVo vo = new OrgVo();
                    BeanUtil.copyProperties(e, vo);
                    return vo;
                }).collect(Collectors.toList());
        IteratorUtils.iterator(result, list
                , (v, p) -> Objects.equals(v.getId(), p.getOrgId())
                , p -> PojoConvertor.toVo(OrgVo.class, p)
                , OrgVo::setChildren
        );
        return result;
    }
}

File: wlxx-services/wlxx-system/wlxx-system-provider/src/main/kotlin/com/xzg/wlxx/system/service/impl/OrgServiceImpl.kt
package com.xzg.wlxx.system.service.impl

import cn.hutool.core.bean.BeanUtil
import cn.hutool.core.collection.CollUtil
import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl
import com.xzg.wlxx.system.client.entity.po.OrgPo
import com.xzg.wlxx.system.client.entity.vo.OrgVo
import com.xzg.wlxx.system.client.exception.BusinessException
import com.xzg.wlxx.system.mapper.OrgMapper
import com.xzg.wlxx.system.service.OrgService
import lombok.RequiredArgsConstructor
import lombok.extern.slf4j.Slf4j
import org.springframework.stereotype.Service
import java.util.*
import java.util.function.Consumer
import java.util.stream.Collectors

@Service
@Slf4j
@RequiredArgsConstructor
class OrgServiceImpl : ServiceImpl<OrgMapper?, OrgPo?>(), OrgService {
    override fun add(org: OrgPo): Boolean {
        val isRepeat = ktQuery().eq(OrgPo::orgCode, org.orgCode)
            .exists()
        if (isRepeat) {
            throw BusinessException("orgCode is repeat")
        }
        return save(org)
    }

    override fun list4Tree(rootId: Long?): List<OrgVo?> {
        val list = ktQuery().orderByAsc(OrgPo::orgLevel)
            .list()
        val rootLevel: Int? = list[0]?.orgLevel
        val result = list.stream()
            .filter { e: OrgPo? -> if (Objects.nonNull(rootId)) e?.id == rootId else e?.orgLevel == rootLevel }
            .map<OrgVo?> { e: OrgPo? ->
                val vo = OrgVo()
                BeanUtil.copyProperties(e, vo)
                vo
            }.collect(Collectors.toList<OrgVo?>())
        iterator(list, result)
        return result
    }

    fun iterator(list: List<OrgPo?>, result: List<OrgVo?>) {
        result.forEach(Consumer<OrgVo?> { e: OrgVo? ->
            val children = list.stream()
                .filter { sub: OrgPo? -> e?.id == sub?.orgId }
                .map<OrgVo?> { map: OrgPo? ->
                    val vo = OrgVo()
                    BeanUtil.copyProperties(map, vo)
                    vo
                }.collect(Collectors.toList<OrgVo?>())
            if (!CollUtil.isEmpty(children)) {
                iterator(list, children)
                
            }
        })
    }
}
